src/ExtendedSystem.cpp

```diff
@@ -40,84 +40,171 @@ void ExtendedSystem::Init(PDBSetup &pdb, Velocity &vel,  config_setup::Input inp
+      stRange = numAtomsInBox[0];
+      endRange = stRange + numAtomsInBox[1] - 1;
+    } else {
+      std::cout << "Error: Only Box 0 and Box 1 supported!" << std::endl;
+      exit(EXIT_FAILURE);
+    }
+
+    XYZArray binaryCoorSOA(binaryCoor);
+
+    pdb.atoms.min[b].x = *std::min_element(binaryCoorSOA.x + stRange, binaryCoorSOA.x + endRange);
+    pdb.atoms.min[b].y = *std::min_element(binaryCoorSOA.y + stRange, binaryCoorSOA.y + endRange);
+    pdb.atoms.min[b].z = *std::min_element(binaryCoorSOA.z + stRange, binaryCoorSOA.z + endRange);
+    pdb.atoms.max[b].x = *std::max_element(binaryCoorSOA.x + stRange, binaryCoorSOA.x + endRange);
+    pdb.atoms.max[b].y = *std::max_element(binaryCoorSOA.y + stRange, binaryCoorSOA.y + endRange);
+    pdb.atoms.max[b].z = *std::max_element(binaryCoorSOA.z + stRange, binaryCoorSOA.z + endRange);
+  }
 }
 
-void ExtendedSystem::UpdateVelocity(PDBSetup &pdb, Velocity &vel, const char *filename, const int box,
-                                    MoleculeLookup & molLookup, Molecules & mols, int & cmIndex)
-{
-  // We must read restart PDB, which hold correct
-  // number atom info in each Box
-  int numAtoms = pdb.atoms.numAtomsInBox[box];
-  int moleculeOffset = 0;
-  XYZ *binaryVel;
-  binaryVel = new XYZ[numAtoms];
-  read_binary_file(filename, binaryVel, numAtoms);
-  //find the starting index
 
-  for(; cmIndex < (int) molLookup.molLookupCount; cmIndex++) {
-    if(moleculeOffset >= numAtoms) break;
-    int currentMolecule = molLookup.molLookup[cmIndex];
-    int numberOfAtoms = mols.start[currentMolecule + 1] - mols.start[currentMolecule];
-    int atomDestinationStart = mols.start[currentMolecule];
 
-    for(int atom = 0; atom < numberOfAtoms; atom++) {
-      vel.x[atomDestinationStart + atom] = binaryVel[moleculeOffset + atom].x;
-      vel.y[atomDestinationStart + atom] = binaryVel[moleculeOffset + atom].y;
-      vel.z[atomDestinationStart + atom] = binaryVel[moleculeOffset + atom].z;
+void ExtendedSystem::ReadVelocity(PDBSetup &pdb, config_setup::Input inputFiles, MoleculeLookup & molLookup,
+                                     Molecules & mols){
+    // We must read restart PDB, which hold correct
+  // number atom info in each Box
+  // We define local variables here, for case that
+  // Binary Velocity is provided but not binary coordinates.
+  int numAtoms = 0;
+  int boxStart[BOX_TOTAL];
+  int numAtomsInBox[BOX_TOTAL];
+  boxStart[0] = 0;
+  for(int b = 0; b < BOX_TOTAL; b++) {
+    if(inputFiles.files.binaryVelInput.defined[b]) {
+      if (mols.restartFromCheckpoint){
+        numAtomsInBox[b] = molLookup.restartedNumAtomsInBox[b];
+      } else {
+        numAtomsInBox[b] = pdb.atoms.numAtomsInBox[b];
+      }
     }
+  }
 
-    moleculeOffset += numberOfAtoms;
+  for(int b = 0; b < BOX_TOTAL; b++) {
+    if(inputFiles.files.binaryVelInput.defined[b]) {
+      numAtoms += numAtomsInBox[b];
+      if (b == 1)
+        boxStart[1] = numAtomsInBox[0];
+    }
```

Comment:
Can't you make it more general to support simulation with more than 2 boxes?

Reply:
This calculation is done once in PDBAtoms and reused for binary files.  Also, it is now general for any number of boxes.